### ISSUE TITLE
UUID/logical type removal

### ISSUE THREAD

**Issue by sjk83** [fastavro/fastavro#341]
I am trying to use fastavro in a restricted environment where syscalls to clone are not allowed. On import the python uuid module runs a shell command which calls clone a number of times.  

I have two questions:
1. I am new to avro so the use of logical types is not clear to me. The avro documentation mentions logical types, but I can't find a uuid logical type. Where did this come from? what is it used for?

2. will removing the uuid logical type have any side effects? Will this only prevent me from using the uuid logical type and nothing else? I have a hacked together version of fast avro that does this, but some of the tests are failing outside of just things that are UUID related.

Thanks!

**Comment by scottbelden** [comment]
> I am new to avro so the use of logical types is not clear to me. The avro documentation mentions logical types, but I can't find a uuid logical type. Where did this come from? what is it used for?

The `UUID` logical type was added in the last 1.9.0 avro specification: https://avro.apache.org/docs/current/spec.html#UUID

As for the use of logical types in general, the idea is probably easiest explained by thinking about something like a `Date`. If you want to use a date in a schema, you might be tempted to use a string with a format like `DD-MM-YYYY`. However, the main problem with this is that any consumer of the record needs to know the format of that string. If a consumer thinks the format is `MM-DD-YYYY` then when they parse it they will either get the wrong date or probably some exception. The way avro solved this was that a date could be specified as an `int` number of days since 1 January 1970 and client libraries (if supported) would transform the `int` into a native date type (on Python, that's a `datetime` object). This standardizes the format of date representation.

> will removing the uuid logical type have any side effects? Will this only prevent me from using the uuid logical type and nothing else? I have a hacked together version of fast avro that does this, but some of the tests are failing outside of just things that are UUID related.

Yes, it should have no other side effects than the fact that it won't create a `UUID` object in python. There is a related issue about having [a way to disable logical type conversions](https://github.com/fastavro/fastavro/issues/116) which I think would also solve this. I have a branch where I played around with that some but never found an API I felt comfortable with.

**Comment by sjk83** [comment]
> The UUID logical type was added in the last 1.9.0 avro specification: https://avro.apache.org/docs/current/spec.html#UUID
Thanks for the link and the quick response! That is very helpful, I will read through that at length after this. 

Your explanation on logical types makes a lot of sense. It seems like a good feature to have, and potentially one I will want/need to use!

> There is a related issue about having a way to disable logical type conversions which I think would also solve this. 
I saw this issue before I posted, but the solution won't work for my purposes. The issue isn't anywhere in fastavro that calls the uuid module functions, or in the uuid functions themselves. The issue is on import of the uuid module. 

> Yes, it should have no other side effects than the fact that it won't create a UUID object in python.
I think I can live with that. I will take another crack at removing the uuid imports and getting the tests to pass.

**Comment by scottbelden** [comment]
Oh, I re-read the original post and that sounds frustrating. It seems like in your case it might be better to build a version of python that has a modified `uuid` module rather than having to fork all the libraries you use. Of course, that's up to you to decide which would require more effort.

I'm going to close this issue unless there's anything else to discuss. Feel free to continue posting here if you have trouble fixing the tests in the fork.

**Comment by sjk83** [comment]
>  It seems like in your case it might be better to build a version of python that has a modified uuid module rather than having to fork all the libraries you use.
I did think about this. For now fastavro is the only module using uuid, so it is likely easier to modify fastavro.

I got rid of the uuid module and got the tests to pass. i added some extra assertions to check that the behavior is what I expect. Everything seems fine.

The uuids are now encoded and decoded as strings with no checking if they conform to RFC-4122 (from the avro spec). Is this what you would expect upon removal of all UUID() objects? 

Thanks again for your help!
